Refactor Trino onboarding and add more questions. (#25)
883G/Onboarding-Newbies: chapter_06/day_20.md

   - Understand how Trino plans and optimizes queries for efficient execution across distributed data sources.
   - Explore query optimization techniques such as predicate pushdown, join reordering, and cost-based optimization.

#### Chapter 3: Advanced Features and Optimization

- **Topics Covered:**
  - **Spill To Disk:**
    - Explanation of the spill-to-disk mechanism, its role in handling large queries, and scenarios where it is triggered.
  - **Dynamic Filtering:**
    - Overview of dynamic filtering, including its role in optimizing query execution by reducing scanned data.
  - **Object Storage:**
    - Understanding the integration of Trino with object storage systems such as S3, GCS, and other Storage.
  - **Functions & Operators:**
    - Insights into combining functions to create complex expressions for advanced querying.
  - **User Defined Functions:**
    - Overview of UDFs in Trino, including how to create and use custom functions to extend Trino's capabilities.

#### Core Concepts

##### 1. **Spill To Disk:**

>     - Explanation of Trino’s spill-to-disk mechanism for handling large queries that exceed memory limits.

Reviewer: not that clear, what are you excpecting the "student" to answer? i woudv'e preferd something like Explain how trino spills data to disk? when it os heppninig? 
or something like that.

    - Configuration options and performance considerations.
    - Understand scenarios where spill-to-disk is triggered and how it ensures query reliability.

##### 2. **Dynamic Filtering:**
    - Understanding how dynamic filtering improves query efficiency by reducing scanned data.
    - Use cases and best practices for enabling dynamic filtering.
    - Understand how dynamic filtering works with distributed data sources to optimize resource utilization.

##### 3. **Object Storage:**

>     - Learn about Trino’s integration with object storage systems like S3, GCS, and Azure Blob Storage.

Reviewer: - Learn about Trino’s integration with object storage systems like S3, GCS, Azure Blob Storage,HDFS, and etc. student may not think about our hdfs :(

17) How does Trino integrate with Hive metastore, and why is this integration significant?
18) How does Trino differ from ETL tools in its approach to data integration?
19) How does Trino secure query execution and access to data sources?
20) What authentication and authorization mechanisms are supported by Trino?
21) What is Trino Gateway, and how does it enhance Trino's routing and workload management capabilities?
22) How does Trino Gateway handle query routing across multiple Trino clusters, and what benefits does this provide in multi-cluster deployments?
23) What are the main downsides or limitations of using Trino?
(Consider factors like query performance at extreme scale, compatibility with certain data sources, operational complexity, or limitations in handling highly transactional workloads.)

## **Wrapping Up:** :hourglass_flowing_sand:
Reflect on today's learning's with your mentor and peers. Discuss potential projects or use cases where you can apply Trino for distributed query processing and analytics. Consider how Trino can enhance your data analysis capabilities and streamline your data workflows.

## Recommended Articles and Videos:

> - [Trino (PrestoSQL) Documentation](https://trino.io/docs/current/index.html) - Official documentation to explore Trino's features, architecture, and best practices.

Reviewer: i woudv'e added trino Official youtube channel

> - [Trino (PrestoSQL) Documentation](https://trino.io/docs/current/index.html) - Official documentation to explore Trino's features, architecture, and best practices.

Reviewer: a small task like quering a table using code/superset/datagrip might be helpful for fuder understending.

2) Can you explain how Trino executes queries across different data sources without moving data?
3) What are the key features of Trino? what it does better than Spark, Hive and Impala?
4) Which data sources are supported by Trino, and how is connectivity achieved?
5) What is the role of the coordinator and worker in Trino's distributed query execution model?
6) How does Trino achieve high performance and efficiency in query processing?
7) Can you describe a use case where Trino would be particularly beneficial?
8) What is the role of a catalog in Trino? and how is it configured?
9) How does Trino handle query retries or failovers to ensure reliability?
10) Compare Trino to Hive,Impala and spark refer to query complexity, use case suitability,performance and more.
11) What is
12) How does Trino optimize queries involving joins or aggregations across large datasets?
13) How does Trino handle data types from different data sources?
14) How does Trino differ from traditional databases (Postgres, MySQL)?
15) What are Trino Resource Groups, and how are they used for workload management?
16) What is the difference between Trino and Presto?

> 17) How does Trino integrate with Hive metastore, and why is this integration significant?

Reviewer 2: ```suggestion
17) How does Trino integrate with the Hive metastore, and why is this integration significant?
```

18) How does Trino differ from ETL tools in its approach to data integration?
19) How does Trino secure query execution and access to data sources?
20) What authentication and authorization mechanisms are supported by Trino?

> 21) What is Trino Gateway, and how does it enhance Trino's routing and workload management capabilities?

Reviewer 2: ```suggestion
21) What is the Trino Gateway, and how does it enhance Trino's routing and workload management capabilities?
```

> 1) What is Apache Trino and what problem does it solve?
> 2) Can you explain how Trino executes queries across different data sources without moving data?
> 3) What are the key features of Trino? what it does better than Spark, Hive and Impala?
> 4) Which data sources are supported by Trino, and how is connectivity achieved?
> 5) What is the role of the coordinator and worker in Trino's distributed query execution model?
> 6) How does Trino achieve high performance and efficiency in query processing?
> 7) Can you describe a use case where Trino would be particularly beneficial?
> 8) What is the role of a catalog in Trino? and how is it configured?
> 9) How does Trino handle query retries or failovers to ensure reliability?
> 10) Compare Trino to Hive,Impala and spark refer to query complexity, use case suitability,performance and more.
> 11) What is
> 12) How does Trino optimize queries involving joins or aggregations across large datasets?
> 13) How does Trino handle data types from different data sources?
> 14) How does Trino differ from traditional databases (Postgres, MySQL)?
> 15) What are Trino Resource Groups, and how are they used for workload management?
> 16) What is the difference between Trino and Presto?
> 17) How does Trino integrate with Hive metastore, and why is this integration significant?
> 18) How does Trino differ from ETL tools in its approach to data integration?
> 19) How does Trino secure query execution and access to data sources?
> 20) What authentication and authorization mechanisms are supported by Trino?
> 21) What is Trino Gateway, and how does it enhance Trino's routing and workload management capabilities?
> 22) How does Trino Gateway handle query routing across multiple Trino clusters, and what benefits does this provide in multi-cluster deployments?
> 23) What are the main downsides or limitations of using Trino?
> (Consider factors like query performance at extreme scale, compatibility with certain data sources, operational complexity, or limitations in handling highly transactional workloads.)

Reviewer 2: ```suggestion
1) What is Apache Trino and what problem does it solve?
2) Explain how Trino executes queries across different data sources without moving data?
3) What are the key features of Trino? How does it differ from Spark, Hive and Impala? In which scenarios would i recomend to use Trino?
4) Which data sources are supported by Trino, and how is connectivity achieved?
5) What is the role of the coordinator and worker in Trino's distributed query execution model?
6) How does Trino achieve high performance and efficiency in query processing?
7) Can you describe a use case where Trino would be particularly beneficial?
8) What is the role of a catalog in Trino? and how is it configured?
9) How does Trino handle query retries or failovers to ensure reliability?
10) Compare Trino to Hive,Impala and spark refer to query complexity, use case suitability,performance and more.
11) How does Trino optimize queries involving joins or aggregations across large datasets?
12) How does Trino handle data types from different data sources?
13) How does Trino differ from traditional databases (Postgres, MySQL)?
14) What are Trino Resource Groups, and how are they used for workload management?
15) What is the difference between Trino and Presto?
16) How does Trino integrate with Hive metastore, and why is this integration significant?
17) How does Trino differ from ETL tools in its approach to data integration?
18) How does Trino secure query execution and access to data sources?
19) What authentication and authorization mechanisms are supported by Trino?
20) What is Trino Gateway, and how does it enhance Trino's routing and workload management capabilities?
21) How does Trino Gateway handle query routing across multiple Trino clusters, and what benefits does this provide in multi-cluster deployments?
22) What are the main downsides or limitations of using Trino?
(Consider factors like query performance at extreme scale, compatibility with certain data sources, operational complexity, or limitations in handling highly transactional workloads.)
```

Author: Done

## **Wrapping Up:** :hourglass_flowing_sand:
Reflect on today's learning's with your mentor and peers. Discuss potential projects or use cases where you can apply Trino for distributed query processing and analytics. Consider how Trino can enhance your data analysis capabilities and streamline your data workflows.

## Recommended Articles and Videos:

> - [Trino Documentation](https://trino.io/docs/current/index.html) - Official documentation to explore Trino's features, architecture, and best practices.
> - [Trino: The Definitive Guide](https://dokumen.pub/trino-the-definitive-guide-sql-at-any-scale-on-any-storage-in-any-environment-2nbsped-109813723x-9781098137236.html) - A comprehensive guide to mastering Trino for distributed query processing and analytics.
> - [Trino: An Origin Story](https://www.youtube.com/c/trinodb) - Trino official youtube channel.

Reviewer 2: ```suggestion
- [Trino Documentation](https://trino.io/docs/current/index.html) - Official documentation to explore Trino's features, architecture, and best practices.
- [Trino: The Definitive Guide](https://dokumen.pub/trino-the-definitive-guide-sql-at-any-scale-on-any-storage-in-any-environment-2nbsped-109813723x-9781098137236.html) - A comprehensive guide to mastering Trino for distributed query processing and analytics.
- [Trino: An Origin Story](https://www.youtube.com/watch?v=_VUQ-Jh-M68) - A short into video.
-[Trino official youtube channel] (https://www.youtube.com/c/trinodb) - Trino official youtube channel.
```
